internal/config/tokens.go

```diff
@@ -0,0 +1,364 @@
+	"github.com/superfly/flyctl/internal/logger"
+	"github.com/superfly/flyctl/internal/task"
+	"github.com/superfly/macaroon"
+	"github.com/superfly/macaroon/flyio"
+	"golang.org/x/exp/maps"
+)
+
+// UserURLCallback is a function that opens a URL in the user's browser. This is
+// used for token discharge flows that require user interaction.
+type UserURLCallback func(ctx context.Context, url string) error
+
+// MonitorTokens does some housekeeping on the provided tokens. Then, in a
+// goroutine, it continues to keep the tokens updated and fresh. The call to
+// MonitorTokens will return as soon as the tokens are ready for use and the
+// background job will run until the context is cancelled. Token updates include
+//   - Keeping the tokens synced with the config file.
+//   - Refreshing any expired discharge tokens.
+//   - Pruning expired or invalid token.
+//   - Fetching macaroons for any organizations the user has been added to.
+//   - Pruning tokens for organizations the user is no longer a member of.
+func MonitorTokens(monitorCtx context.Context, t *tokens.Tokens, uucb UserURLCallback) {
+	log := logger.FromContext(monitorCtx)
+
+	updated1, err := fetchOrgTokens(monitorCtx, t)
+	if err != nil {
+		log.Debugf("failed to fetch missing tokens for user: %s", err)
+	}
+
+	updated2, err := refreshDischargeTokens(monitorCtx, t, uucb)
+	if err != nil {
+		log.Debugf("failed to update discharge tokens: %s", err)
+
+	}
+
+	if t.FromConfigFile != "" && updated1 || updated2 {
+		if err := SetAccessToken(t.FromConfigFile, t.All()); err != nil {
+			log.Debugf("failed to persist updated tokens: %s", err)
+		}
+	}
+
+	task.FromContext(monitorCtx).Run(func(taskCtx context.Context) {
+		taskCtx, cancelTask := context.WithCancel(taskCtx)
+
+		var m sync.Mutex
+		var wg sync.WaitGroup
+
+		wg.Add(2)
+
+		if t.FromConfigFile != "" {
+			log.Debugf("monitoring tokens at %s", t.FromConfigFile)
+		} else {
+			log.Debug("monitoring tokens in memory")
+		}
+
+		go monitorConfigTokenChanges(taskCtx, &m, t, wg.Done)
+		go keepConfigTokensFresh(taskCtx, &m, t, uucb, wg.Done)
+
+		// shut down when the task manager is shutting down or when the
+		// ctx passed into MonitorTokens is cancelled.
+		select {
+		case <-taskCtx.Done():
+		case <-monitorCtx.Done():
+		}
+
+		log.Debug("done monitoring tokens")
+		cancelTask()
+		wg.Wait()
+	})
+}
+
+// monitorConfigTokenChanges watches for token changes in the config file. This can
+// happen if a foreground process updates the config file while the agent is
+// running.
+func monitorConfigTokenChanges(ctx context.Context, m *sync.Mutex, t *tokens.Tokens, done func()) error {
+	defer done()
+
+	fromConfigFile := tokensConfigFile(t, m)
+	if fromConfigFile == "" {
+		return nil
+	}
+
+	ticker := time.NewTicker(15 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			currentStr, err := ReadAccessToken(t.FromConfigFile)
```

Comment:
why is `t.FromConfigFile` accessed directly here but under lock (`tokensConfigFile()`) earlier?

Reply:
I need to make another fly-go change to put all the Tokens struct fields behind accessor functions that use a mutex (I'll go do that now). The mutex that is used elsewhere in this file is created in this function and avoids concurrent access by the two goroutines that this file launches.

Reply:
Doing this in https://github.com/superfly/fly-go/pull/35 and 94f715a89cc9871f0442e714f5877705e7d95470